lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,476 @@
+		AtomicTxnsInnerUtxoOps: innerUtxoOps,
+	})
+
+	return utxoOpsForTxn, totalInput, totalOutput, totalFees, nil
+}
+
+func _verifyAtomicTxnsWrapper(txn *MsgDeSoTxn) error {
+	// Since the wrapper does not require a public key nor a corresponding signature, we force both
+	// the transaction public key to be the ZeroPublicKey and the signature to be nil.
+	if !NewPublicKey(txn.PublicKey).IsZeroPublicKey() {
+		return RuleErrorAtomicTxnsWrapperPublicKeyMustBeZero
+	}
+	if txn.Signature.Sign != nil {
+		return RuleErrorAtomicTxnsWrapperSignatureMustBeNil
+	}
+
+	// We force TxInputs on the wrapper to be empty for several reasons:
+	//	(1) This is consistent with the logic found in _connectBasicTransferWithExtraSpend()
+	//		that forces TxInputs to be empty following the balance model fork.
+	//	(2) Allowing TxInputs to not be empty would lead to an attack vector were the transaction
+	//		size may be bloated with random TxInputs that do nothing.
+	//	(3) Leads to consistent hashing for the same atomic transaction wrapper and its inner transactions.
+	//	(4) It's generally safer to be more restrictive on the transaction structure.
+	if len(txn.TxInputs) != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroInputs
+	}
+
+	// We force TxOutputs on the wrapper to be empty even though this field is still used post balance model fork.
+	// The reason is this transaction is effectively "signed" by the ZeroPublicKey which is a potential
+	// burn address. Hence, allowing TxOutputs to be populated would enable un-burning DESO which we do not want.
+	if len(txn.TxOutputs) != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroOutputs
+	}
+
+	// There exists three design options for txn.TxnFeeNanos rules in atomic transaction wrappers:
+	//	(1) Force txn.TxnFeeNanos to equal zero.
+	//	(2) Force txn.TxnFeeNanos to equal the sum of the internal transaction's txn.TxnFeeNanos fields.
+	//	(3) Ignore txn.TxnFeeNanos entirely.
+	//
+	// Because txn.TxnFeeNanos gets used in several places for non-connection logic (e.g. BMF),
+	// it's important to use design option (2) to be consistent across core. This check as a result
+	// becomes extremely important in _connectAtomicTxns().
+	var totalInnerTxnFees uint64
+	var err error
+	for _, innerTxn := range txn.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns {
+		totalInnerTxnFees, err = SafeUint64().Add(totalInnerTxnFees, innerTxn.TxnFeeNanos)
+		if err != nil {
+			return RuleErrorAtomicTxnsWrapperHasInternalFeeOverflow
+		}
+	}
+	if txn.TxnFeeNanos != totalInnerTxnFees {
+		return RuleErrorAtomicTxnsWrapperMustHaveEqualFeeToInternalTxns
+	}
+
+	// Technically, the txn.TxnNonce field could be
+	if txn.TxnNonce.ExpirationBlockHeight != 0 || txn.TxnNonce.PartialID != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroedNonce
+	}
+
+	// Since the wrapper is free and modifiable by anyone, we check to ensure the
+	// associated ExtraData is empty to prevent free storage on the blockchain.
```

Comment:
I'm not sure we want this. If we're charging for the wrapper it's fine and could be useful?